### DISCUSSION TITLE
Refactoring components and redesign

### DISCUSSION THREAD

**Discussion by yangwao** [kodadot/nft-gallery#3710] [General]
Hello, so we've kicked redesign as mentioned earlier on #3644  

![image](https://user-images.githubusercontent.com/5887929/187194831-6b06c570-04f2-4ccc-a152-b1a42e6f3d52.png)

Yeah, it's getting there and will be really lovely refreshment for KodaDot

# What do we plan to do in the first phase?
We decided to go with easy steps. Easy. 😅
We would like to redesign the whole K with an iterative rolling phase. That means we will try to get things out into production once it's finished on the design pipeline and iterate with various things in codebase as fast we can do, sounds courageous and brave 😅 

# Namely, we will go in a particular order
1. Landing page + Search+Searchbar - #3844 
1. Gallery item - #4111 
1. Explorer v0.9 - #4378
1. Collection page - 
1. Create / Minting - #4752
1. Explorer v2 -
1. Profile 
1. Stats tables 
1. Others  

# Batch of newly redesigned small components 
- Navbar - https://github.com/kodadot/nft-gallery/issues/3946
- Actions component
- https://github.com/kodadot/nft-gallery/issues/4119
- #4252
- #4143
- #4405 
- #4751 
- #4635 
- #4593 
- #4403 
- #2837 
- #614 

# What needs to be done, and how do we desire to proceed?
The thing is that lot of our components are copypasta, which is not good, and we should strive for highly functional and reusable components as we have more multichain stuff incoming.

As we feel special and love to break things often, we finally decided to migrate from Vue2.7.x to Vue3. 
**But, the thing is Buefy doesn't support Vue3 and Oruga will drop Vue2 support, soon** 
So after chatting with jtommy (author of buefy & oruga) we thought it would result in double migration. Well what we can do if we want to do revamp without doing an extensive code freeze, avoid backporting upgrades to new code, avoid any extra bugs &c and keep history (most precious).
Though once refactor for particular components will be done, we can start making upgrades components to Oruga.
As Oruga is a perfect CSS agnostic framework, we however continue with Bulma as the majority of our contributors understand it well which I bet help us do some small adjustments per se.
The good stuff is that I bet components we will support current needed stuff in Oruga in current state so no need worry about Oruga dropping Vue2, on the contrary, it will push us to migrate much sooner to Vue3.

The good news is that Nuxt3 is coming along as well. 
With Nuxt3 and Vue3, **we can cut build times with Vite and remove transpilling lot of packages.** 
Fast builds on Netlify, faster builds at your hands locally, faster hot-reload, development would feel much more smooth as back then again.

If I did not make any mistake in my "calculations", it should work very well, and our current contributors will be super happy for speedy updates on their local machines.

I hope the whole support of the code review guild will pinpoint during refactoring what we can do and how we can leverage the writing of components.

## Steps to be done
- Lookup components gradually we will need for redesign
  - landing
  - gallery item + collection page
  - explorer
- Make issues for particular refactors
  - we have actually ones in under [Refactoring label](https://github.com/kodadot/nft-gallery/issues?q=is%3Aopen+is%3Aissue+label%3Arefactoring)
- Landing page refactor
  - including components on Landing page  
- Gallery item + collection refactor
  - beware of various logic we are using there for shopping utils, please be cautious to not break 🥲
- Introduce Oruga with new design
  - Landing page and components
  - Gallery item + collection detail
- And so on with other components ... 
- Switch to Vue3 with some adjustments. (HARD one I bet)
- Switching to Nuxt3 (This will be a total surprise 😄)

## Who is involved 
Right now point of contact for design is @exezbcz, on Discord is Exez. 
We decided that @preschian can help us with refactoring components as we noticed is really good on stuff with help of @Jarsen136, but I bet it's better to be done this by one person to not break some mental models

## Design
- Figma link is coming soon. One particular design is done

### Other CSS frameworks considered 
- [Tailwind](https://github.com/tailwindlabs/tailwindcss)
- Chakra
- Windy CSS
- [Unocss](https://github.com/unocss/unocss)

<img src="https://media0.giphy.com/media/w7dn7xRSHGZUs/giphy.gif"/>

## Conclusion
I'm looking forward to the new revamp getting rolling and removing a lot of burdens we currently have 🥳
Feel free to drop any questions you might have. I fairly estimated that the whole process could be completed in under eight weeks, I bet we can do it under six weeks 🤞😁
Once done, it would be the biggest gift to the KodaDot community and Polkadot ecosystem out there.

### Ref
- https://github.com/buefy/buefy
- https://github.com/oruga-ui/oruga
- https://github.com/oruga-ui/theme-bulma

**Comment by yangwao** [comment]
We will try to merge this PR https://github.com/kodadot/nft-gallery/pull/2947 as we talked with @exezbcz that having a default white background would be required to roll new design partially, and @roiLeo introduced some good polishes there anyway! :)

Then all white will come
- https://github.com/kodadot/nft-gallery/issues/3713

**Comment by prachi00** [comment]
Have we started working on this yet? would love to work on a page and revamp the design 😊

**Comment by preschian** [comment]
Hi guys, after a little chit-chat with @yangwao. how do we continue with a redesign landing page **without blocking the `main` branch**?

I'm thinking of using [feature toggles](https://martinfowler.com/articles/feature-toggles.html) on query parameters. So, the new design only can be viewed behind this URL: `kodadot.xyz?newDesign=true`. we still keep and don't break the existing design until the redesign on the landing page finished

for example, on the landing page, modify this file https://github.com/kodadot/nft-gallery/blob/main/pages/index.vue#L1-L3 to something like this:
```vue
<template>
  <Landing v-if="newDesign" />
  <LandingOld v-else />
</template>
```

`newDesign` will be provided by Vue3 `composable`. these rules will apply to another component also in case needed. another example, on the `Footer` component:
```vue
<template>
  <Footer v-if="newDesign" />
  <FooterOld v-else />
</template>
```

with this approach, we can safely merge the redesign components into the `main` branch. after the landing page is finished, we need to clean up `old` components 

what do you guys think?
maybe anyone has a better approach?

**Comment by yangwao** [comment]
As discussed here https://github.com/kodadot/nft-gallery/pull/3895

We are running a few things for "upgrade" 

1. Visual redesign to a new style of Landing page at `main` branch
2. Switching to Vue3 with composition API at `canary` branch
3. Moving towards Oruga, component library 
4. Switching to Nuxt3 without Nitro Bridge 
5. Drop webpack and migrate to Vite https://github.com/kodadot/nft-gallery/issues/3908
6. Merging things from `canary` to `main` (this one will have epic amount of issues)

Tough one!  

As discussed with @vikiival we would like to go with upgrading Vue3, which would mean a new branch, `canary` 
Canary will be experimental, and once ready, we will merge things into `main` 
Canary will be okay if things sometimes won't work as it would be thought one, but also would be nice if things would work before merging into `main`
In canary we will have only new components, so we can filter out old stuff and make it as opportunity to refactor things into small components and introduce more test suites where it matters 

We will keep current flow merging things into stable `main` (beta.kodadot.xyz) -> `hyper-jpeg` (kodadot.xyz)
`canary` will have a new Netlify deployment 
- #3909 

The issue why we are doing this is that **we have a ready design for landing page on the hand**, for gallery and collection detail is in the works and the rest will definitely take a while, at least 3-6 weeks, plus some week two on top implementing stuff there.

Thus moving fast, I would like to see gradually see this implemented as we still have a small audience (<10k users daily) than anticipated. As I believe UX would hugely benefit just from landing already. 

To address concerns https://github.com/kodadot/nft-gallery/pull/3895#issuecomment-1235295502
We will proceed just with the landing page with this as the rest will take definitely a while and will open room for marketing efforts.

**Reply by preschian** [reply]
thanks for the explanation

> Thus moving fast, I would like to see gradually see this implemented as we still have a small audience (<10k users daily) than anticipated. As I believe UX would hugely benefit just from landing already.

I just want to make sure this one. Does it mean we can merge redesigned components in the main branch without [query parameters approach](https://github.com/kodadot/nft-gallery/discussions/3710#discussioncomment-3501133)? if yes, we can close this PR https://github.com/kodadot/nft-gallery/pull/3895 and start redesign components gradually

**Reply by yangwao** [reply]
Good Q! 
I would keep with query parameters for now till the whole landing is done, so in any case, we would need to push something out to the production, it won't be blocked. Cherry picking is not much preferred as it might accidentally break things.

**Comment by shashkovdanil** [comment]
Guys, what do you think about migration from `Buefy` to self-written components? I mean, `Buefy` is good for MVP, but it's hard to customize

**Reply by roiLeo** [reply]
New components are based on [oruga ui](https://oruga.io/) combined with [histoire](https://histoire.dev/) stories (#4167).

> `Buefy` is good for MVP, but it's hard to customize

I don't really agree with that, since the code base was not made by one person, I find it easier to maintain than having handmade components.

**Reply by vikiival** [reply]
Agree with @roiLeo, customizing Oruga is much easier than doing everything by hand,

+ Buefy and Oruga, share almost same API so the migration should be smoother

**Comment by yangwao** [comment]
Added minting and section with other minor but important components we brought to light

**Comment by yangwao** [comment]
Thanks for all the fish

<img src="https://media4.giphy.com/media/7eVp9MHlNI90c/giphy.gif"/>

- https://github.com/kodadot/nft-gallery/pull/7583
